### PR TITLE
Start lnd in e2e tests

### PR DESCRIPTION
**Pull request by D4nte**
The output of this PR:
- lnds are started for the actors
- lnd wallets are initialised
- RPC clients are ready
- lnds public keys are printed in logs

helps with #1951

### CODE REVIEW

File: api_tests/lib/bitcoind_instance.ts [hunk 1]
```diff
@@ -18,39 +18,25 @@ export class BitcoindInstance implements LedgerInstance {
         private readonly logDir: string,
         public readonly p2pPort: number,
         public readonly rpcPort: number
-    ) {}
+    ) {
+        this.dbDir = tmp.dirSync();
```
Comment (thomaseizinger) [inline api_tests/lib/bitcoind_instance.ts]: We are mixing sync and async code here. If you need an async constructor, you can use the factory method pattern and have a static async fn on this class.

File: api_tests/lib_sdk/lnd_instance.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+            "regtest",
+            "admin.macaroon"
+        );
+    }
+
+    public getGrpcSocket() {
+        return "127.0.0.1:" + this.actorConfig.lndRpcPort;
+    }
+
+    private writeLogFile(bitcoindDataDir: string) {
```
Comment (thomaseizinger) [inline api_tests/lib_sdk/lnd_instance.ts]: More of a config file, no? :)

File: api_tests/lib_sdk/utils.ts [hunk 3]
```diff
@@ -83,3 +84,11 @@ export async function createDefaultSwapRequest(counterParty: Actor) {
     };
     return swapRequest;
 }
+
+export async function fileExists(filepath: string) {
+    let logFileExists = false;
+    do {
+        await sleep(500);
+        logFileExists = fs.existsSync(filepath);
```
Comment (thomaseizinger) [inline api_tests/lib_sdk/utils.ts]: Please don't mix sync and async code.

File: api_tests/lib_sdk/utils.ts [hunk 4]
```diff
@@ -83,3 +84,11 @@ export async function createDefaultSwapRequest(counterParty: Actor) {
     };
     return swapRequest;
 }
+
+export async function fileExists(filepath: string) {
```
Comment (thomaseizinger) [inline api_tests/lib_sdk/utils.ts]: Would be good if the fn name reflects the waiting until the file actually exists.

File: api_tests/lib_sdk/lnd_instance.ts [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        this.lndDir = this.logDir + "/lnd-" + this.actorConfig.name;
+        fs.mkdirSync(this.lndDir);
+
+        this.writeLogFile(bitcoindDataDir);
+    }
+
+    public async start() {
+        const bin = process.env.LND_BIN ? process.env.LND_BIN : "lnd";
+
+        if (global.verbose) {
```
Comment (thomaseizinger) [inline api_tests/lib_sdk/lnd_instance.ts]: You have a logger accessible here, no need to use the legacy logging.

File: api_tests/lib/ledger_runner.ts [hunk 6]
```diff
@@ -45,6 +45,7 @@ export class LedgerRunner {
                         await getPort({ port: 18444 }),
                         await getPort({ port: 18443 })
                     );
+                    global.bitcoind = instance;
```
Comment (thomaseizinger) [inline api_tests/lib/ledger_runner.ts]: Why can we not use existing way of accessing data from the bitcoind instance on the global object?

File: api_tests/lib/bitcoind_instance.ts [hunk 7]
```diff
@@ -78,4 +64,31 @@ export class BitcoindInstance implements LedgerInstance {
...
+rpcallowip=0.0.0.0/0
+nodebug=1
+rest=1
+acceptnonstdtxn=0
+zmqpubrawblock=tcp://127.0.0.1:28332
+zmqpubrawtx=tcp://127.0.0.1:28333
+`;
+        const config = this.dbDir.name + "/bitcoin.conf";
+
+        fs.writeFile(config, output, function(err: any) {
```
Comment (thomaseizinger) [inline api_tests/lib/bitcoind_instance.ts]: Any reason for deviating from using async/await here?

File: api_tests/lib/bitcoind_instance.ts [hunk 8]
```diff
@@ -78,4 +64,31 @@ export class BitcoindInstance implements LedgerInstance {
...
+        const output = `regtest=1
+server=1
+printtoconsole=1
+bind=0.0.0.0:${this.p2pPort}
+rpcbind=0.0.0.0:${this.rpcPort}
+rpcallowip=0.0.0.0/0
+nodebug=1
+rest=1
+acceptnonstdtxn=0
+zmqpubrawblock=tcp://127.0.0.1:28332
```
Comment (thomaseizinger) [inline api_tests/lib/bitcoind_instance.ts]: Why are these ports hardcoded?

File: api_tests/lib_sdk/lnd_instance.ts [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+    private lndGrpc: any;
+
+    constructor(
+        private readonly logger: Logger,
+        private readonly logDir: string,
+        private readonly actorConfig: E2ETestActorConfig,
+        bitcoindDataDir: string
+    ) {
+        this.lndDir = this.logDir + "/lnd-" + this.actorConfig.name;
+        fs.mkdirSync(this.lndDir);
```
Comment (thomaseizinger) [inline api_tests/lib_sdk/lnd_instance.ts]: Please don't mix async and sync calls. It shouldn't matter much in this particular case but it is a mental burden to reading "sync" and having to think if it matters.

File: api_tests/lib_sdk/utils.ts (hunk 4 above)
Comment (D4nte) [reply api_tests/lib_sdk/utils.ts]: Haha @tcharding

**Comment by D4nte** [comment]
@tcharding this is not mergeable until we provide a Docker/bin of Lnd. We might want to do that in a separate PR and then rebase.

**Comment by D4nte** [comment]
@tcharding I assigned to you to do @thomaseizinger's comments. Please note that most of the `fs` async function are `callback` based. You may want to use `promisify` to make them return a `Promise` to allow you to use the `await`/`async` keywords.

Also, I think docker might be the best for `lnd`.

Let me know if you need help with either of those.

File: api_tests/lib_sdk/utils.ts (hunk 4 above)
Comment (tcharding) [reply api_tests/lib_sdk/utils.ts]: Yep, that one was me :)

Comment (thomaseizinger) [reply api_tests/lib_sdk/utils.ts]: Suggestions: `untilFileExists`, reads nicely with `await`:

```js
await untilFileExists(...)
```

File: api_tests/lib/bitcoind_instance.ts (hunk 7 above)
Comment (tcharding) [reply api_tests/lib/bitcoind_instance.ts]: Until these comments I had paid no attention, and had no idea, that there were async and sync versions of all the file system commands.

Comment (D4nte) [reply api_tests/lib/bitcoind_instance.ts]: > Any reason for deviating from using async/await here?

No reason except that this works so why promisify-ing? This clearly something obvious to you @thomaseizinger but not us.

**Comment by tcharding** [comment]
Ok, its been a massive shit fight today but I believe we have all the concerns and suggestions from this thread of discussion as well as from over on the [original PR](https://github.com/comit-network/comit-rs/pull/1984) this is based on addressed except for one - @bonomat's concerns that the LND stuff should not be in `Artors`.  I have no thoughts on this right now so just flagging it as unresolved.

File: api_tests/lib_sdk/lnd_instance.ts [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+        const { lnd } = lnService.authenticatedLndGrpc({
+            cert,
+            macaroon,
+            socket: this.getGrpcSocket(),
+        });
+
+        this.lndGrpc = lnd;
+        this.logger.debug("Waiting for lnd to catch up with blocks");
+        await this.logReader().waitForLogMessage(
+            "LNWL: Done catching up block hashes"
```
[the remark this answers is not in the document]
Comment (tcharding) [reply api_tests/lib_sdk/lnd_instance.ts]: Its an LND log message, taken straigt from the generated logs.

[the remark this answers is not in the document]
Comment (tcharding) [reply api_tests/lib_sdk/lnd_instance.ts]: No clue what it stands for ...

[the remark this answers is not in the document]
Comment (D4nte) [reply api_tests/lib_sdk/lnd_instance.ts]: Lightning Network WaLlet is my guess. Why does it matter?

**Comment by tcharding** [comment]
```
        await aliceLnd = alice.startLnd();
        await bobLnd = bob.startLnd();

        await aliceLnd.fund();

        await aliceLnd.connect(bobLnd);
        await aliceLnd.openChannel(bobLnd);

        const invoice = await bobLnd.addInvoice(aliceLnd); // ??? Parameter might need to be `bob`.
        await aliceLnd.sendPayment(invoice);

        await aliceLnd.assertInvoiceSettled(invoice);
        await bobLnd.assertInvoiceSettled(invoice);

        await aliceLnd.assertChannelBalanceSender();
        await bobLnd.assertChannelBalanceReceiver();
```

File: api_tests/lib/bitcoind_instance.ts (hunk 8 above)
Comment (tcharding) [reply api_tests/lib/bitcoind_instance.ts]: > 've some concerns that the functions `fund/connectLND/openChannel` are in the wrong place

Please check out [this commit](https://github.com/comit-network/comit-rs/pull/2020/commits/c87ed75a1e4ddb7083e0f0e12da5b2189f960fd8).  It moves all the lnd code out of `Actor` and into a class of its own.

File: api_tests/lib_sdk/lnd_instance.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (bonomat) [reply api_tests/lib_sdk/lnd_instance.ts]: Just wanted to know if it makes sense to wait for this particular log message.

[the remark this answers is not in the document]
Comment (D4nte) [reply api_tests/lib_sdk/lnd_instance.ts]: Please see the `logger.debug` just above `"Waiting for lnd to catch up with blocks"` :)

File: api_tests/lib/config.ts [hunk 11]
```diff
@@ -20,7 +20,10 @@ export class E2ETestActorConfig {
     constructor(
         public readonly httpApiPort: number,
         public readonly comitPort: number,
-        public readonly name: string
+        public readonly name: string,
+        public readonly lndP2pPort: number,
+        public readonly lndRpcPort: number,
+        public readonly lndRestPort: number
```
Comment (D4nte) [inline api_tests/lib/config.ts]: This is not needed.

File: api_tests/lib_sdk/lnd.ts [hunk 12]
```diff
@@ -0,0 +1,202 @@
+import { ChildProcess, spawn } from "child_process";
+import * as fs from "fs";
+import { E2ETestActorConfig } from "../lib/config";
+import { waitUntilFileExists } from "./utils";
+import * as path from "path";
+import lnService from "ln-service";
+import { Logger } from "log4js";
+import { LogReader } from "../lib/log_reader";
+import { mkdirAsync } from "../lib_sdk/utils";
```
Comment (D4nte) [inline api_tests/lib_sdk/lnd.ts]:
```suggestion
import { mkdirAsync } from "./utils";
```

File: api_tests/lib_sdk/lnd.ts [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+debuglevel=debug
+
+; peer to peer port
+listen=127.0.0.1:${this.actorConfig.lndP2pPort}
+
+; gRPC
+rpclisten=127.0.0.1:${this.actorConfig.lndRpcPort}
+
+; REST interface
+restlisten=127.0.0.1:${this.actorConfig.lndRestPort}
```
Comment (D4nte) [inline api_tests/lib_sdk/lnd.ts]: This is not needed

File: api_tests/lib/config.ts (hunk 11 above)
Comment (tcharding) [reply api_tests/lib/config.ts]: Why?  If we don't have these then LND defaults to its own default ports - these will then clash between the two instances, let alone when we try to parallelise the tests.

File: api_tests/lib/ledger_runner.ts (hunk 6 above)
Comment (tcharding) [reply api_tests/lib/ledger_runner.ts]: Can you extrapolate please, I don't see how to?

File: api_tests/lib/config.ts (hunk 11 above)
Comment (D4nte) [reply api_tests/lib/config.ts]: So let's deactivate it then,

Comment (tcharding) [reply api_tests/lib/config.ts]: Oh, you are talking about just the rest port.  You are right it isn't needed.

File: api_tests/lib_sdk/lnd_instance.ts (hunk 2 above)
Comment (tcharding) [reply api_tests/lib_sdk/lnd_instance.ts]: Changed to `createConfigFile`

File: api_tests/lib_sdk/utils.ts (hunk 4 above)
Comment (tcharding) [reply api_tests/lib_sdk/utils.ts]: Changed to `waitUntilFileExists` as @D4nte  originally had it :)

File: api_tests/lib_sdk/lnd_instance.ts (hunk 5 above)
Comment (tcharding) [reply api_tests/lib_sdk/lnd_instance.ts]: Fixed as suggested.

File: api_tests/lib/bitcoind_instance.ts (hunk 8 above)
Comment (tcharding) [reply api_tests/lib/bitcoind_instance.ts]: Moved to the constructor, as done for the other ports.

File: api_tests/lib_sdk/actors/actor.ts [hunk 14]
```diff
@@ -518,6 +529,56 @@ export class Actor {
...
+        await sleep(1);
+    }
+
+    public async connectLnd(other: Actor) {
+        await other.dummy();
+    }
+
+    public async openChannel(other: Actor) {
+        await other.dummy();
+    }
```
[the remark this answers is not in the document]
Comment (tcharding) [reply api_tests/lib_sdk/actors/actor.ts]: Moved code out of `Actor` -> `lnd.ts`.

File: api_tests/lib_sdk/lnd_instance.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (tcharding) [reply api_tests/lib_sdk/lnd_instance.ts]: Moved to `start()`.

File: api_tests/lib_sdk/lnd.ts (hunk 12 above)
Comment (tcharding) [reply api_tests/lib_sdk/lnd.ts]: Done, thanks.

File: api_tests/lib_sdk/lnd.ts (hunk 13 above)
Comment (tcharding) [reply api_tests/lib_sdk/lnd.ts]: Fixed as discussed offline.

File: api_tests/lib/ledger_runner.ts (hunk 6 above)
Comment (tcharding) [reply api_tests/lib/ledger_runner.ts]: @thomaseizinger I'm going to leave this for a follow up PR, this is not trivial to resolve unless I'm missing something.

Comment (D4nte) [reply api_tests/lib/ledger_runner.ts]: I'll tackle this once #2015 is done.

**Comment by tcharding** [comment]
bors+

**Comment by D4nte** [comment]
Not sure what is going on with bors.

bors r+

File: api_tests/lib/ledger_runner.ts (hunk 6 above)
Comment (thomaseizinger) [reply api_tests/lib/ledger_runner.ts]: What I meant was: we already expose the bitcoind instance through global.ledgers.bitcoin

**Pull request merged**